File: src/leetcode/python/SerializeAndDeserializeBinaryTree_297.py
class Index(object):
    def __init__(self, x):
        self.index = x


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class SerializeAndDeserializeBinaryTree_297(object):
    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        def serial(node):
            if node is None:
                ser.append("$,")
                return
            ser.append(str(node.val) + ",")
            serial(node.left)
            serial(node.right)

        ser = []
        serial(root)
        return ''.join(ser)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        def deserial():
            if i.index >= len(sd) or i.index < len(sd) and sd[i.index] == "$":
                return None
            node = TreeNode(int(sd[i.index]))
            i.index += 1
            node.left = deserial()
            i.index += 1
            node.right = deserial()
            return node

        sd = data.split(",")
        i = Index(0)
        return deserial()

File: src/leetcode/java/SerializeAndDeserializeBinaryTree_297.java
package leetcode.java;

public class SerializeAndDeserializeBinaryTree_297 {
  private static final String NULL = "$";

  // Encodes a tree to a single string.
  public String serialize(TreeNode root) {
    StringBuilder builder = new StringBuilder();
    serialize(root, builder);
    return builder.toString();
  }

  private void serialize(TreeNode root, StringBuilder builder) {
    if (root == null) {
      builder.append(NULL + ",");
      return;
    }
    builder.append(root.val + ",");
    serialize(root.left, builder);
    serialize(root.right, builder);
  }

  // Decodes your encoded data to tree.
  public TreeNode deserialize(String data) {
    String[] dchar = data.split(",");
    return deserialize(dchar, new Index(0));
  }

  private TreeNode deserialize(String[] data, Index i) {
    if (i.index >= data.length || i.index < data.length && data[i.index].equals(NULL)) return null;
    TreeNode node = new TreeNode(Integer.valueOf(data[i.index]));
    i.index++;
    node.left = deserialize(data, i);
    i.index++;
    node.right = deserialize(data, i);
    return node;
  }

  class Index {
    int index;
    public Index(int i) { this.index = i; }
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
